source/yats/include/yats/scheduler.h

```diff
@@ -31,60 +41,88 @@ class scheduler
 
     void run()
     {
-        std::vector<abstract_task_container*> to_run;
-        for (auto& elem : m_tasks)
+        if (!initial_schedule())
+        {
+            return;
+        }
+
+        while (auto guard = m_condition.wait_main(thread_group::main_thread_number()))
         {
-            to_run.push_back(elem.get());
+            auto current_task = get(thread_group::main_thread_number());
+            m_tasks[current_task]->run();
+
+            // This can be removed after the change of control flow.
+            schedule_following(current_task);
         }
+    }
 
-        std::function<void()> check_runnable;
+protected:
+    size_t get(size_t constraint)
+    {
+        std::unique_lock<std::mutex> guard(m_mutex);
+        auto current_task = m_tasks_to_process[constraint].front();
+        m_tasks_to_process[constraint].pop();
+        return current_task;
+    }
+
+    void schedule(size_t index)
+    {
+        const auto& constraints = m_tasks[index]->constraints();
 
-        check_runnable = [&to_run, &check_runnable, this]() {
-            std::lock_guard<std::mutex> lock(m_mutex);
+        // If there are multiple threads we can chose from take the one with the smallest current workload
+        // TODO: theres probably a better method to check which thread to take
+        auto constraint_it = std::min_element(constraints.cbegin(), constraints.cend(), [this](size_t lhs, size_t rhs) {
+            return m_tasks_to_process[lhs].size() < m_tasks_to_process[rhs].size();
+        });
 
-            while (true)
+        m_tasks_to_process[*constraint_it].push(index);
+        m_condition.notify(*constraint_it);
+    }
+
+    void schedule_following(size_t index)
+    {
+        std::unique_lock<std::mutex> guard(m_mutex);
+        for (auto next_task : m_tasks[index]->following_nodes())
+        {
+            if (m_tasks[next_task]->can_run())
             {
-                auto runnable = std::find_if(to_run.begin(), to_run.end(), [](abstract_task_container* task) {
-                    return task->can_run();
-                });
-
-                if (runnable == to_run.end())
-                {
-                    return;
-                }
-                auto task = *runnable;
-                to_run.erase(runnable);
-
-                m_threads.emplace_back([&to_run, &check_runnable, task]() {
-                    task->run();
-                    check_runnable();
-                });
+                schedule(next_task);
             }
-        };
+        }
+    }
 
-        check_runnable();
-        while (true)
+    bool initial_schedule()
```

Comment:
Please consider renaming the function. schedule_initial_task might be more fitting. Having the method retrurn a value is debatable. Throwing an exception might be better.
We also discussed that initial_schedule should save the initial tasks for subsequent calls to run.

Reply:
Exceptions are typically used for when we encounter an error. Here the return value simply indicates whether any tasks were scheduled at all.

Reply:
But we must not return false.